### DISCUSSION TITLE
potential uniforms plugin

### DISCUSSION THREAD

**Discussion by myk002** [DFHack/dfhack#1943] [Ideas]
Hi,

I have some ideas about a potential `uniforms` plugin. I notice that we don't really have a lot of tools that manipulate uniforms, and it occurs to me that it might be because working with uniforms is hard and I should ask for feedback before I dig myself into a hole (channel?).

I've looked at [gui/clone-uniform](https://github.com/DFHack/scripts/blob/master/gui/clone-uniform.lua), [gui/choose-weapons](https://github.com/DFHack/scripts/blob/master/gui/choose-weapons.lua), and [uniform-unstick](https://github.com/DFHack/scripts/blob/master/uniform-unstick.lua) to get the general idea of what data is available and how it is currently being used by scripts.

The primary use case I have in mind is similar to the [orders](https://docs.dfhack.org/en/latest/docs/Plugins.html#orders) plugin - export an existing uniform to JSON using string identifiers in the raws and import it later in a different fort. This is essentially what is requested in #651. My particular interest is reliably automating the setup of dreamfort. Since elements of uniforms vary by civ, modifications to uniforms can't be easily scripted with a quickfort blueprint. I plan to publish a set of pre-made uniforms that players can import in their onMapLoad.init files (via something like `once-per-save uniforms import melee,ranged`). Once it is automated for dreamfort, it can also be used by any fort.

There are lots of things I don't know. Is it safe to overwrite existing uniforms or must the plugin always create a new uniform on import (`clone-uniform` creates new uniforms so I assume creating uniforms is safe)? How difficult is it to match a uniform element's properties (e.g. "metal") to the raws and how hard is it to look up on import? Is the uniform definition directly used by the units that are "wearing" that uniform or is the definition copied to the units at "apply uniform" time? These are questions I plan to investigate more, but if anyone has bits of knowledge (or warnings) at their fingertips, hearing them now would help me.

**Comment by se5a** [comment]
Would this potentially allow the edit of "invisible" uniforms used for eg miners, & tree fellers?

**Comment by ab9rf** [comment]
i'm wondering if we need, or could at least use, a DSL to specify how to adaptively react to elements which might or might not be available to a civ
